### ISSUE TITLE
QUIC support

### ISSUE THREAD

**Issue by AlexLeung** [denoland/deno#1558]
It would be nice if the Deno standard library had QUIC protocol support out-of-the-box, implemented in Rust underneath. QUIC will likely be the UDP-based alternative to TCP used as the foundation for the Internet Engineering Task Force (IETF)'s upcoming [HTTP3 spec](https://tools.ietf.org/html/draft-ietf-quic-http-17). Essentially QUIC is a more efficient implementation of the combined capabilities of TCP and TLS, and a less mature version of QUIC has been used by Google for years in requests from Chrome to its own webservers. Node.js has [an open issue](https://github.com/nodejs/node/issues/23064) for adding in QUIC support, but based on the discussion there I'm assuming that the priority is low for them. They may not understand the significance of adding support.

**Comment by ry** [comment]
I agree - would be great - and seems relatively accessible with
https://github.com/djc/quinn
I’m open to anyone who wants to attempt this. And if someone doesn’t pick it up, I would try myself once a few other things have been taken care of (We still don’t have basic TLS support for example- that should come first)

**Comment by haoxinst** [comment]
Maybe we can also consider https://github.com/cloudflare/quiche

**Comment by trivikr** [comment]
Existing implementations of QUIC: https://github.com/quicwg/base-drafts/wiki/Implementations

The ones (under active development) in Rust:
* [quinn](https://github.com/djc/quinn)
* [quiche](https://github.com/cloudflare/quiche)

**Comment by 95th** [comment]
`quiche` seems to expose only mio based non-blocking API and not an async API.

**Comment by cretz** [comment]
From my experience with quiche, it's only the CPU bound parts, you bring your own network IO. I have used it with a Tokio UDP socket and converted it to a modern Rust async futures stream+sink. The library is basically only the hard parts of state machine, parsing, and crypto pieces.

The bigger problem with using quiche is that they require BoringSSL and have multiple times stated they are unlikely to change (see https://github.com/cloudflare/quiche/issues/129#issuecomment-524035205 and https://github.com/cloudflare/quiche/issues/16#issuecomment-460205108). But there is an [old PR open yet on hold](https://github.com/cloudflare/quiche/pull/126) for OpenSSL so maybe there's hope for rustls.

**Comment by humancalico** [comment]
@ry Is anyone working on this issue? I could give this a shot. Can't give any guarantees though.

**Comment by kitsonk** [comment]
Make sure you check out #3995 as it is more comprehensive of an issue (I missed this one when I opened that) and Ry had some thoughts on implementation.

**Comment by fimbault** [comment]
An empirical study on QUIC implementations that's worth reading https://qlog.edm.uhasselt.be/epiq
And a summary if you don't have much time https://qlog.edm.uhasselt.be/epiq/images/summary.png

**Comment by kitsonk** [comment]
Even though #3995 came after this one, let's close it in lieu of #3995 as it is more comprehensive.

**Comment by balupton** [comment]
Other links that may be of interest:

- https://en.wikipedia.org/wiki/HTTP/3
- https://github.com/cloudflare/quiche
- https://web.dev/webtransport/
- https://github.com/w3c/webtransport

**Comment by waldyrious** [comment]
For reference, this issue has been reopened following @ry's [comment](https://github.com/denoland/deno/issues/3995#issuecomment-823376063) when closing #3995 (which this issue had been closed as a duplicate/subset of):

> HTTP/2 shipped in v1.9.0 (see #9935)
>
> HTTP/3 and QUIC can be discussed in a new issue.

**Comment by rektide** [comment]
hyper underpins [deno's native http extensions]( https://github.com/denoland/deno/blob/main/ext/http/lib.rs). there is hyperium/hyper#1818 for http3 support in that framework.

**Comment by ahmadkakarr** [comment]
when http/3 can be implemented in deno?

**Comment by trusktr** [comment]
What's interesting about HTTP/3 is that people will make ESM CDN servers that can server-push module graphs to clients and possibly beat bundlers, and this highly aligns with Deno's vanilla-esm-and-importmaps philosophy. The sooner we have HTTP/3, the sooner we might have even better vanilla ESM support in Deno (and others) where module downloads will be very optimized (f.e. all gzipped and cached from the CDN).

**Comment by devsnek** [comment]
i'm working on an initial implementation of QUIC here https://github.com/denoland/deno/pull/21942

**Comment by masx200** [comment]
https://github.com/caddyserver/caddy/issues/5086

**Comment by masx200** [comment]
Many of Cloudflare's websites have been blocked.

any update?
I found that the https TLS connection of the existing website was attacked by a man-in-the-middle, and the connection was disconnected, and the connection was successful only by using quic "xxxxxxxxxxx.xxxxxxxxxxxxxxxxx.xxxxxxxxx.workers.dev"!

我发现有的网站的https的tls连接被中间人攻击了,连接被断开了,只有使用quic才能连接成功!比如说"xxxxxxxxxxx.xxxxxxxxxxxxxxxxx.xxxxxxxxx.workers.dev".

At present, browsers Firefox, Chrome, and Edge already support HTTP3, and there is also a query of HTTPS type through DNS, and you can find out the server's support for HTTP3.

目前浏览器firefox,chrome,edge都已经支持http3了,还有通过dns查询https类型,可以查到服务器对于http3的支持情况.

I found that Chrome was able to open Cloudflare's workers.dev website, but caddy couldn't access workers.dev.

**Comment by higlyrecommand** [comment]
@masx200 

> Many of Cloudflare's websites have been blocked.
> 
> any update? I found that the https TLS connection of the existing website was attacked by a man-in-the-middle, and the connection was disconnected, and the connection was successful only by using quic "xxxxxxxxxxx.xxxxxxxxxxxxxxxxx.xxxxxxxxx.workers.dev"!
> 
> 我发现有的网站的https的tls连接被中间人攻击了,连接被断开了,只有使用quic才能连接成功!比如说"xxxxxxxxxxx.xxxxxxxxxxxxxxxxx.xxxxxxxxx.workers.dev".
> 
> At present, browsers Firefox, Chrome, and Edge already support HTTP3, and there is also a query of HTTPS type through DNS, and you can find out the server's support for HTTP3.
> 
> 目前浏览器firefox,chrome,edge都已经支持http3了,还有通过dns查询https类型,可以查到服务器对于http3的支持情况.
> 
> I found that Chrome was able to open Cloudflare's workers.dev website, but deno couldn't access workers.dev.

Your issue is not related to `Deno`, as it is caused by the Great Firewall of China. The Great Firewall is an IDS (Intrusion Detection System) bypass detector that can reassemble TCP packets and then perform struct decode and matching. Additionally, TLS handshake packets are not encrypted, and these packets contain SNI (Server Name Indication) information. The Great Firewall can easily identify and block subsequent connections. Moreover, I can tell you that the Initial Packet of HTTP/3 (QUIC) is also not encrypted, so the Great Firewall can potentially recognize and block HTTP/3 connections in future updates.

**Comment by SerJaimeLannister** [comment]
Any updates on this?

**Comment by o-t-w** [comment]
All web browsers now support HTTP/3 so support in Deno would be great. https://caniuse.com/http3

**Comment by CyanChanges** [comment]
This was merged which implemented QUIC API in Deno.
#21942

**Comment by SerJaimeLannister** [comment]
Yes this is awesome , any way of accessing this right now in deno since I am excited in trying out the quic api.


Also since you've supported quic , does your api also support http/3 ?

and can I use this directly with the deno.serve -ish?
```
Deno.serve((_req) => {
  return new Response("Hello, World!");
});
``

like I would love it , if we could simply replace it directly with another function or  a cli flag Deno.serve_http3 or maybe

deno run --http3  main.ts ? 

or something like this.


This is really good and I thank you for building this from my side.

Hope that you achieve the best in the world!

**Comment by CyanChanges** [comment]
> Yes this is awesome , any way of accessing this right now in deno since I am excited in trying out the quic api.
> 
> Also since you've supported quic , does your api also support http/3 ?
> 
> and can I use this directly with the deno.serve -ish?
> 
> ```
> Deno.serve((_req) => {
>   return new Response("Hello, World!");
> });
> ``
> 
> like I would love it , if we could simply replace it directly with another function or  a cli flag Deno.serve_http3 or maybe
> 
> deno run --http3  main.ts ? 
> 
> or something like this.
> 
> 
> This is really good and I thank you for building this from my side.
> 
> Hope that you achieve the best in the world!
> ```

I guess there's only something like `Deno.openQuic`
There's no HTTP/3 support yet ig?

**Comment by SerJaimeLannister** [comment]
How hard is http/3 support though?

**Comment by unixfox** [comment]
@SerJaimeLannister please do not spam the users subscribed to this issue.

Only the QUIC interface has been merged, now there are still work in progress to be able to use "fetch" using HTTP3. You can follow the work here: https://github.com/denoland/deno/pull/27431

Once there will be something concrete I think it will be announced here.

**Comment by devsnek** [comment]
HTTP/3 is on the 2025 roadmap for hyper (the http library used by deno). As that moves along, we will be able to hook it up with the QUIC layer in the various places you would expect it to be.

**Comment by talentlessguy** [comment]
h3 (hyper's custom QUIC implementation) seems to support both servers and clients, but it's not available in Hyper yet

https://github.com/hyperium/h3/blob/master/examples/server.rs

**Comment by taikulawo** [comment]
You can try https://github.com/taikulawo/nghttp3. A production ready `nghttp3` rust binding with tokio supports

**Comment by fry69** [comment]
FYI: https://blog.cloudflare.com/async-quic-and-http-3-made-easy-tokio-quiche-is-now-open-source/

> A little over 6 years ago, we presented [quiche](https://blog.cloudflare.com/enjoy-a-slice-of-quic-and-rust/), our open source QUIC implementation written in Rust. Today we’re announcing the open sourcing of [tokio-quiche](https://crates.io/crates/tokio-quiche), our battle-tested, asynchronous QUIC library combining both quiche and the Rust Tokio async runtime. Powering Cloudflare’s Proxy B in Apple iCloud Private Relay and our next-generation [Oxy-based](https://blog.cloudflare.com/introducing-oxy/) proxies, tokio-quiche handles millions of HTTP/3 requests per second with low latency and high throughput. tokio-quiche also powers [Cloudflare Warp’s MASQUE](https://blog.cloudflare.com/zero-trust-warp-with-a-masque/) client, replacing our WireGuard tunnels with QUIC-based tunnels, and the async version of [h3i](https://blog.cloudflare.com/h3i/).

**Comment by bartlomieju** [comment]
The QUIC portion of this request has shipped. Deno now has a full QUIC API, `Deno.connectQuic`, `QuicEndpoint`, `QuicListener`, `QuicConn`, unidirectional/bidirectional streams, and 0-RTT (currently behind `--unstable`), landed in #21942, and WebTransport is built on top of it (#27431).

What's not yet done is HTTP/3 for `fetch` and `Deno.serve`. That's gated on HTTP/3 support in [hyper](https://github.com/hyperium/hyper) (the HTTP library Deno uses), which is on hyper's 2025 roadmap via [`hyperium/h3`](https://github.com/hyperium/h3). Once that lands we'll be able to hook it into the QUIC layer in the usual places.

Since this issue (opened in 2019) bundled QUIC and HTTP/3 together and the QUIC API is now in place, we're closing it as the umbrella request. The remaining HTTP/3 work is tracked separately in #34712 so it stays focused. Thanks to everyone for the input and links over the years.
